From 5e696381e3bb695efe1570a55f2f5f0137fc9c13 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Brice=20Andr=C3=A9?= <brice.andre@transport.alstom.com>
Date: Sun, 15 Jun 2014 15:12:12 +0200
Subject: [PATCH] Corrected a bug in wxWidgets preventing integration of
 accents in wxRichText if they were recorded from old versions of wxWidgets
 with non-unicode encoding

---
 wxWidgets3/include/wx/richtext/richtextbuffer.h | 10 +++---
 wxWidgets3/include/wx/richtext/richtextxml.h    |  2 +-
 wxWidgets3/src/richtext/richtextxml.cpp         | 46 ++++++++++++++++---------
 3 files changed, 35 insertions(+), 23 deletions(-)

diff --git a/wxWidgets3/include/wx/richtext/richtextbuffer.h b/wxWidgets3/include/wx/richtext/richtextbuffer.h
index 2cf636a..93f2273 100644
--- a/wxWidgets3/include/wx/richtext/richtextbuffer.h
+++ b/wxWidgets3/include/wx/richtext/richtextbuffer.h
@@ -2395,7 +2395,7 @@ public:
     /**
         Imports this object from XML.
     */
-    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse);
+    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding);
 #endif
 
 #if wxRICHTEXT_HAVE_DIRECT_OUTPUT
@@ -2967,7 +2967,7 @@ public:
     virtual wxString GetTextForRange(const wxRichTextRange& range) const;
 
 #if wxUSE_XML
-    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse);
+    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding);
 #endif
 
 #if wxRICHTEXT_HAVE_DIRECT_OUTPUT
@@ -4430,7 +4430,7 @@ public:
     virtual bool UsesParagraphAttributes() const { return false; }
 
 #if wxUSE_XML
-    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse);
+    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding);
 #endif
 
 #if wxRICHTEXT_HAVE_DIRECT_OUTPUT
@@ -4707,7 +4707,7 @@ public:
     virtual bool UsesParagraphAttributes() const { return false; }
 
 #if wxUSE_XML
-    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse);
+    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding);
 #endif
 
 #if wxRICHTEXT_HAVE_DIRECT_OUTPUT
@@ -5704,7 +5704,7 @@ public:
     virtual wxString GetTextForRange(const wxRichTextRange& range) const;
 
 #if wxUSE_XML
-    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse);
+    virtual bool ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding);
 #endif
 
 #if wxRICHTEXT_HAVE_DIRECT_OUTPUT
diff --git a/wxWidgets3/include/wx/richtext/richtextxml.h b/wxWidgets3/include/wx/richtext/richtextxml.h
index 013aa97..dec29f1 100644
--- a/wxWidgets3/include/wx/richtext/richtextxml.h
+++ b/wxWidgets3/include/wx/richtext/richtextxml.h
@@ -190,7 +190,7 @@ public:
 #endif
 
     /// Recursively import an object
-    bool ImportXML(wxRichTextBuffer* buffer, wxRichTextObject* obj, wxXmlNode* node);
+    bool ImportXML(wxRichTextBuffer* buffer, wxRichTextObject* obj, wxXmlNode* node, const wxString& original_encoding);
 #endif
 
     /// Creates an object given an XML element name
diff --git a/wxWidgets3/src/richtext/richtextxml.cpp b/wxWidgets3/src/richtext/richtextxml.cpp
index cf413f4..2d359ed 100644
--- a/wxWidgets3/src/richtext/richtextxml.cpp
+++ b/wxWidgets3/src/richtext/richtextxml.cpp
@@ -108,7 +108,7 @@ bool wxRichTextXMLHandler::DoLoadFile(wxRichTextBuffer *buffer, wxInputStream& s
                     {
                     }
                     else
-                        ImportXML(buffer, buffer, child);
+                        ImportXML(buffer, buffer, child, xmlDoc->GetFileEncoding());
                 }
 
                 child = child->GetNext();
@@ -139,10 +139,10 @@ wxRichTextObject* wxRichTextXMLHandler::CreateObjectForXMLName(wxRichTextObject*
 }
 
 /// Recursively import an object
-bool wxRichTextXMLHandler::ImportXML(wxRichTextBuffer* buffer, wxRichTextObject* obj, wxXmlNode* node)
+bool wxRichTextXMLHandler::ImportXML(wxRichTextBuffer* buffer, wxRichTextObject* obj, wxXmlNode* node, const wxString& original_encoding)
 {
     bool recurse = false;
-    obj->ImportFromXML(buffer, node, this, & recurse);
+    obj->ImportFromXML(buffer, node, this, & recurse, original_encoding);
 
     // TODO: how to control whether to import children.
 
@@ -158,7 +158,7 @@ bool wxRichTextXMLHandler::ImportXML(wxRichTextBuffer* buffer, wxRichTextObject*
                 if (childObj)
                 {
                     compositeParent->AppendChild(childObj);
-                    ImportXML(buffer, childObj, child);
+                    ImportXML(buffer, childObj, child, original_encoding);
                 }
             }
             child = child->GetNext();
@@ -344,7 +344,7 @@ bool wxRichTextXMLHandler::ExportXML(wxXmlNode* parent, wxRichTextObject& obj)
     // wxUSE_STREAMS
 
 // Import this object from XML
-bool wxRichTextObject::ImportFromXML(wxRichTextBuffer* WXUNUSED(buffer), wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse)
+bool wxRichTextObject::ImportFromXML(wxRichTextBuffer* WXUNUSED(buffer), wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& WXUNUSED(original_encoding))
 {
     handler->GetHelper().ImportProperties(GetProperties(), node);
     handler->GetHelper().ImportStyle(GetAttributes(), node, UsesParagraphAttributes());
@@ -419,9 +419,9 @@ bool wxRichTextObject::ExportXML(wxXmlNode* parent, wxRichTextXMLHandler* handle
 #endif
 
 // Import this object from XML
-bool wxRichTextPlainText::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse)
+bool wxRichTextPlainText::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding)
 {
-    wxRichTextObject::ImportFromXML(buffer, node, handler, recurse);
+    wxRichTextObject::ImportFromXML(buffer, node, handler, recurse, original_encoding);
 
     if (node->GetName() == wxT("text"))
     {
@@ -466,9 +466,21 @@ bool wxRichTextPlainText::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* nod
             textChild = textChild->GetNext();
         }
 
-        wxString actualText;
-        actualText << (wxChar) wxAtoi(text);
-        SetText(actualText);
+        if (original_encoding.CmpNoCase("UTF-8") == 0)
+        {
+           wxString actualText;
+           actualText << (wxChar) wxAtoi(text);
+           SetText(actualText);
+        }
+        else
+        {
+           char data[2];
+           data[0] = wxAtoi(text);
+           data[1] = 0;
+
+           wxString actualText(data, wxCSConv(original_encoding));
+           SetText(actualText);
+        }
     }
     else
         return false;
@@ -680,9 +692,9 @@ bool wxRichTextPlainText::ExportXML(wxXmlNode* parent, wxRichTextXMLHandler* han
 #endif
 
 // Import this object from XML
-bool wxRichTextImage::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse)
+bool wxRichTextImage::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding)
 {
-    wxRichTextObject::ImportFromXML(buffer, node, handler, recurse);
+    wxRichTextObject::ImportFromXML(buffer, node, handler, recurse, original_encoding);
 
     wxBitmapType imageType = wxBITMAP_TYPE_PNG;
     wxString value = node->GetAttribute(wxT("imagetype"), wxEmptyString);
@@ -830,9 +842,9 @@ bool wxRichTextImage::ExportXML(wxXmlNode* parent, wxRichTextXMLHandler* handler
 #endif
 
 // Import this object from XML
-bool wxRichTextParagraphLayoutBox::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse)
+bool wxRichTextParagraphLayoutBox::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding)
 {
-    wxRichTextObject::ImportFromXML(buffer, node, handler, recurse);
+    wxRichTextObject::ImportFromXML(buffer, node, handler, recurse, original_encoding);
 
     *recurse = true;
 
@@ -924,9 +936,9 @@ bool wxRichTextParagraphLayoutBox::ExportXML(wxXmlNode* parent, wxRichTextXMLHan
 #endif
 
 // Import this object from XML
-bool wxRichTextTable::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse)
+bool wxRichTextTable::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, wxRichTextXMLHandler* handler, bool* recurse, const wxString& original_encoding)
 {
-    wxRichTextBox::ImportFromXML(buffer, node, handler, recurse);
+    wxRichTextBox::ImportFromXML(buffer, node, handler, recurse, original_encoding);
 
     *recurse = false;
 
@@ -940,7 +952,7 @@ bool wxRichTextTable::ImportFromXML(wxRichTextBuffer* buffer, wxXmlNode* node, w
         if (childObj)
         {
             AppendChild(childObj);
-            handler->ImportXML(buffer, childObj, child);
+            handler->ImportXML(buffer, childObj, child, original_encoding);
         }
         child = child->GetNext();
     }
-- 
1.8.4.msysgit.0

